### PR TITLE
fix(iOS): Ensure reports on disk prior to app termination

### PR DESCRIPTION
**Pull request by kattrali**
## Goal

(Extends #287)

Ensure report delivery while holding the native JS execution thread for as little time as possible in the case of non-fatal errors being reported. Updated the integration test to make the stack trace ridiculous.

## Tests

* Tested against physical devices as well as in the integration suite

### Linked issues

Closes #287

- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: cocoa/BugsnagReactNative.m [hunk 1]
```diff
@@ -141,7 +141,9 @@ + (void)startWithConfiguration:(BugsnagConfiguration *)config {
 
 RCT_EXPORT_MODULE()
 
-RCT_EXPORT_METHOD(notify:(NSDictionary *)options) {
+RCT_EXPORT_METHOD(notify:(NSDictionary *)options
+                  resolve:(RCTPromiseResolveBlock)resolve
+                  reject:(RCTPromiseRejectBlock)reject) {
```
Comment (fractalwrench) [inline cocoa/BugsnagReactNative.m]: We should update the declaration in the header file also

File: android/src/main/java/com/bugsnag/BugsnagReactNative.java [hunk 2]
```diff
@@ -107,13 +107,9 @@ public void leaveBreadcrumb(ReadableMap options) {
...
   }
 
-  @ReactMethod
-  public void notify(ReadableMap payload) {
-      notifyBlocking(payload, false, null);
-  }
 
   @ReactMethod
-  public void notifyBlocking(ReadableMap payload, boolean blocking, com.facebook.react.bridge.Callback callback) {
+  public void notify(ReadableMap payload, com.facebook.react.bridge.Promise promise) {
```
Comment (fractalwrench) [inline android/src/main/java/com/bugsnag/BugsnagReactNative.java]: We could get away without the fully qualified class name here - it was necessary previously to disambiguate between the two `Callback` classes

**Comment by kattrali** [comment]
Both nice catches, thanks @fractalwrench !
